community/values/src/main/java/org/neo4j/values/storable/NumberValues.java

```diff
@@ -49,53 +69,74 @@ public static int hash( double number )
             return hash( asLong );
         }
         long bits = Double.doubleToLongBits( number );
-        return (int)(bits ^ (bits >>> 32));
+        return (int) (bits ^ (bits >>> 32));
     }
 
+    /*
+     * This is a slightly silly optimization but by turning the computation
+     * of the hashcode into a dot product we trick the jit compiler to use SIMD
+     * instructions and performance doubles.
+     */
     public static int hash( byte[] values )
```

Comment:
Can we use `Arrays.hashCode(byte[])` and other overloads instead or writing code ourselves?
Presumably they are very optimized and can become intrinsics at some point.

Reply:
They are not intrinsics at the moment and more than twice as slow both in Java 8 and Java 9. When Java10 is on the horizon we should revisit.

Reply:
They do not need a major release to change code or add intrinsics ;)
See https://bugs.openjdk.java.net/browse/JDK-8186203 and https://bugs.openjdk.java.net/browse/JDK-8059113. It just feels strange to me that computing hash code for arrays is our bottleneck.

Reply:
It is probably not a major bottleneck but since our interpreted runtime is basically one big hashmap these things do matter :)

We have jmh-benchmarks set up for hashcodes so we should pick up if `Arrays.hashCode` are made faster all of a sudden. But even so we cannot use them naively every where since we have different equality semantics than that of java. For example, given 

```
ByteArray bytes = Values.byteArray( new byte[]{-42, 42} );
LongArray longs = Values.longArray( new long[]{-42L, 42L} );
```

then in our semantics we expect `bytes.equals(longs) == true` and `bytes.hashCode() == longs.hashCode()`. This is not the case for java equality or java hashcodes.

Reply:
Fair enough, I haven't thought about array equality.